### PR TITLE
refactor: Upgrade the models to use keras 3.0

### PR DESCRIPTION
**Pull request by JGSweets**
This pr:
* refactors the following models to use keras 3.0
    - structured_model
    - unstructured_model
    - CharLoadTFModel
* Fixes a bug in loading a trainable model from library
* updates requirements for keras 3.0
* fixes manifest to not include __pycache__ files

Closes: https://github.com/capitalone/DataProfiler/issues/1126

### CODE REVIEW

File: dataprofiler/labelers/char_load_tf_model.py [hunk 1]
```diff
@@ -237,7 +237,8 @@ def _construct_model(self) -> None:
         model_loc = self._parameters["model_path"]
 
         self._model: tf.keras.Model = tf.keras.models.load_model(model_loc)
-        softmax_output_layer_name = self._model.outputs[0].name.split("/")[0]
+        self._model = tf.keras.Model(self._model.inputs, self._model.outputs)
```
Comment (JGSweets) [inline dataprofiler/labelers/char_load_tf_model.py:240]: Required for the function to have `output_names`. Sequential models do not have that output.

File: dataprofiler/labelers/char_load_tf_model.py [hunk 2]
```diff
@@ -253,20 +254,27 @@ def _construct_model(self) -> None:
...
         losses = {softmax_output_layer_name: "categorical_crossentropy"}
 
         # use f1 score metric
         f1_score_training = labeler_utils.F1Score(
             num_classes=num_labels, average="micro"
         )
-        metrics = {softmax_output_layer_name: ["acc", f1_score_training]}
+        metrics = {
+            softmax_output_layer_name: [
+                "categorical_crossentropy",
```
Comment (JGSweets) [inline dataprofiler/labelers/char_load_tf_model.py:273]: keras v3 requires specification of loss while v2 did not

File: dataprofiler/labelers/char_load_tf_model.py [hunk 3]
```diff
@@ -294,30 +302,33 @@ def _reconstruct_model(self) -> None:
         num_labels = self.num_labels
         default_ind = self.label_mapping[self._parameters["default_label"]]
 
-        # Remove the 2 output layers ('softmax', 'tf_op_layer_ArgMax')
```
Comment (JGSweets) [inline dataprofiler/labelers/char_load_tf_model.py:297]: popping does nothing in v3

File: dataprofiler/labelers/char_load_tf_model.py [hunk 4]
```diff
@@ -294,30 +302,33 @@ def _reconstruct_model(self) -> None:
...
-        for _ in range(2):
-            self._model.layers.pop()
-
         # Add the final Softmax layer to the previous spot
+        # self._model.layers[-2] to skip: original softmax
         final_softmax_layer = tf.keras.layers.Dense(
             num_labels, activation="softmax", name="softmax_output"
-        )(self._model.layers[-4].output)
+        )(self._model.layers[-2].output)
```
Comment (JGSweets) [inline dataprofiler/labelers/char_load_tf_model.py:309]: argmax ops does not show as a layer anymore

File: dataprofiler/labelers/char_load_tf_model.py [hunk 5]
```diff
@@ -294,30 +302,33 @@ def _reconstruct_model(self) -> None:
...
         losses = {softmax_output_layer_name: "categorical_crossentropy"}
 
         # use f1 score metric
         f1_score_training = labeler_utils.F1Score(
             num_classes=num_labels, average="micro"
         )
-        metrics = {softmax_output_layer_name: ["acc", f1_score_training]}
+        metrics = {
+            softmax_output_layer_name: [
+                "categorical_crossentropy",
```
Comment (JGSweets) [inline dataprofiler/labelers/char_load_tf_model.py:327]: keras v3 requires specification of loss while v2 did not

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 6]
```diff
@@ -74,6 +74,133 @@ def create_glove_char(n_dims: int, source_file: str = None) -> None:
             file.write(word + " " + " ".join(str(num) for num in embd) + "\n")
 
 
+@tf.keras.utils.register_keras_serializable(package="CharacterLevelCnnModel")
+class ThreshArgMaxLayer(tf.keras.layers.Layer):
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:78]: Refactored threshargmax out of the class function so it could be properly serialized in keras v3

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 7]
```diff
@@ -74,6 +74,133 @@ def create_glove_char(n_dims: int, source_file: str = None) -> None:
...
+            bg_label_tf,
+            tf.multiply(tf.subtract(argmax_layer, bg_label_tf), argmax_mask),
+            name="ThreshArgMax",
+        )
+        # final_predicted_layer.set_shape(argmax_layer.shape)
+        return final_predicted_layer
+
+
+@tf.keras.utils.register_keras_serializable(package="CharacterLevelCnnModel")
+class EncodingLayer(tf.keras.layers.Layer):
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:150]: as above with the thresh class

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 8]
```diff
@@ -280,7 +407,7 @@ def save_to_disk(self, dirpath: str) -> None:
         labels_dirpath = os.path.join(dirpath, "label_mapping.json")
         with open(labels_dirpath, "w") as fp:
             json.dump(self.label_mapping, fp)
-        self._model.save(os.path.join(dirpath))
+        self._model.save(os.path.join(dirpath, "model.keras"))
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:410]: keras v3 requires the ext to be `.keras`

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 9]
```diff
@@ -301,15 +428,7 @@ def load_from_disk(cls, dirpath: str) -> CharacterLevelCnnModel:
...
-        # use f1 score metric
-        custom_objects = {
-            "F1Score": labeler_utils.F1Score(
-                num_classes=max(label_mapping.values()) + 1, average="micro"
-            ),
-            "CharacterLevelCnnModel": cls,
-        }
-        with tf.keras.utils.custom_object_scope(custom_objects):
-            tf_model = tf.keras.models.load_model(dirpath)
+        tf_model = tf.keras.models.load_model(os.path.join(dirpath, "model.keras"))
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:431]: serialized in keras v3 hence the above custom scopes were not needed

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 10]
```diff
@@ -333,35 +452,6 @@ def load_from_disk(cls, dirpath: str) -> CharacterLevelCnnModel:
         ]
         return loaded_model
 
-    @staticmethod
-    def _char_encoding_layer(
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:337]: moved to a class

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 11]
```diff
@@ -383,47 +473,7 @@ def _argmax_threshold_layer(
         """
         # Initialize the thresholds vector variable and create the threshold
         # matrix.
-        class ThreshArgMaxLayer(tf.keras.layers.Layer):
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:386]: moved to a class

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 12]
```diff
@@ -449,17 +499,13 @@ def _construct_model(self) -> None:
         max_length = self._parameters["max_length"]
         max_char_encoding_id = self._parameters["max_char_encoding_id"]
 
-        # Encoding layer
-        def encoding_function(input_str: tf.Tensor) -> tf.Tensor:
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:453]: moved to a class

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 13]
```diff
@@ -485,7 +530,6 @@ def encoding_function(input_str: tf.Tensor) -> tf.Tensor:
                 max_char_encoding_id + 2,
                 self._parameters["dim_embed"],
                 weights=[embedding_matrix],
-                input_length=input_shape[0],
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:488]: api change in keras v3

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 14]
```diff
@@ -503,7 +547,7 @@ def encoding_function(input_str: tf.Tensor) -> tf.Tensor:
             if self._parameters["dropout"]:
                 self._model.add(tf.keras.layers.Dropout(self._parameters["dropout"]))
             # Add batch normalization, set fused = True for compactness
-            self._model.add(tf.keras.layers.BatchNormalization(fused=False, scale=True))
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:506]: keras v3 api change

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 15]
```diff
@@ -564,22 +614,18 @@ def _reconstruct_model(self) -> None:
         num_labels = self.num_labels
         default_ind = self.label_mapping[self._parameters["default_label"]]
 
-        # Remove the 3 output layers (dense_2', 'tf_op_layer_ArgMax',
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:567]: popping does nothing in keras v3

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 16]
```diff
@@ -564,22 +614,18 @@ def _reconstruct_model(self) -> None:
...
-        #                             'thresh_arg_max_layer')
-        for _ in range(3):
-            self._model.layers.pop()
-
         # Add the final Softmax layer to the previous spot
+        # self._model.layers[-3] to skip: thresh and original softmax
         final_softmax_layer = tf.keras.layers.Dense(
             num_labels, activation="softmax", name="dense_2"
-        )(self._model.layers[-4].output)
+        )(self._model.layers[-3].output)
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:621]: argmax does not show as a layer in v3 hence the reduction

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 17]
```diff
@@ -589,14 +635,20 @@ def _reconstruct_model(self) -> None:
...
+        softmax_output_layer_name = self._model.output_names[0]
         losses = {softmax_output_layer_name: "categorical_crossentropy"}
 
         # use f1 score metric
         f1_score_training = labeler_utils.F1Score(
             num_classes=num_labels, average="micro"
         )
-        metrics = {softmax_output_layer_name: ["acc", f1_score_training]}
+        metrics = {
+            softmax_output_layer_name: [
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:646]: keras v3 requires specification of loss while v2 did not

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 18]
```diff
@@ -729,7 +781,9 @@ def _validate_training(
         for x_val, y_val in val_data:
             y_val_pred.append(
                 self._model.predict(
-                    x_val, batch_size=batch_size_test, verbose=verbose_keras
+                    tf.convert_to_tensor(x_val),
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:784]: v3 requires the conversion to tensor

File: dataprofiler/labelers/data_labelers.py [hunk 19]
```diff
@@ -141,11 +141,11 @@ def load_from_library(cls, name: str, trainable: bool = False) -> BaseDataLabele
         :type trainable: bool
         :return: DataLabeler class
         """
+        for labeler_name, labeler_class_obj in cls.labeler_classes.items():
```
Comment (JGSweets) [inline dataprofiler/labelers/data_labelers.py:144]: fixes a bug in identification

File: dataprofiler/labelers/labeler_utils.py [hunk 20]
```diff
@@ -435,11 +435,6 @@ def get_config(self) -> dict:
         base_config = super().get_config()
         return {**base_config, **config}
 
-    def reset_state(self) -> None:
```
Comment (JGSweets) [inline dataprofiler/labelers/labeler_utils.py:438]: no longer needed as this is builtin

File: dataprofiler/tests/labelers/test_character_level_cnn_model.py [hunk 21]
```diff
@@ -368,9 +371,8 @@ def test_input_encoding(self):
         max_char_encoding_id = 127
         max_len = 10
 
-        encode_output = cnn_model._char_encoding_layer(
-            input_str_tensor, max_char_encoding_id, max_len
-        ).numpy()[0]
+        encode_layer = EncodingLayer(max_char_encoding_id, max_len)
```
Comment (JGSweets) [inline dataprofiler/tests/labelers/test_character_level_cnn_model.py:374]: now tests the encoding layer

File: requirements-ml.txt [hunk 22]
```diff
@@ -1,7 +1,7 @@
 scikit-learn>=0.23.2
-keras>=2.4.3,<3.0.0
+keras>=3.0.0
 rapidfuzz>=2.6.1
-tensorflow>=2.6.4,<2.15.0; sys.platform != 'darwin'
-tensorflow>=2.6.4,<2.15.0; sys_platform == 'darwin' and platform_machine != 'arm64'
-tensorflow-macos>=2.6.4,<2.15.0; sys_platform == 'darwin' and platform_machine == 'arm64'
+tensorflow>=2.16.0; sys.platform != 'darwin'
```
Comment (JGSweets) [inline requirements-ml.txt:4]: update to v3 keras

File: dataprofiler/labelers/labeler_utils.py [hunk 23]
```diff
@@ -358,7 +358,7 @@ def __init__(
 
         def _zero_wt_init(name: str) -> tf.Variable:
             return self.add_weight(
-                name, shape=self.init_shape, initializer="zeros", dtype=self.dtype
+                name=name, shape=self.init_shape, initializer="zeros", dtype=self.dtype
```
Comment (JGSweets) [inline dataprofiler/labelers/labeler_utils.py:361]: the initial error that started this all

Comment (JGSweets) [reply dataprofiler/labelers/labeler_utils.py:361]: keras v3 api change

**Comment by JGSweets** [comment]
Added a commit to drop support for Python 3.8 since tensorflow >= `2.16.0` does not exist.

File: .github/workflows/test-python-package.yml [hunk 24]
```diff
@@ -16,7 +16,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.8, 3.9, "3.10"]
+        python-version: [3.9, "3.10"]
```
Comment (taylorfturner) [inline .github/workflows/test-python-package.yml:19]: nice -- and post merge of this #1090 can add 3.11

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 25]
```diff
@@ -503,7 +547,7 @@ def encoding_function(input_str: tf.Tensor) -> tf.Tensor:
             if self._parameters["dropout"]:
                 self._model.add(tf.keras.layers.Dropout(self._parameters["dropout"]))
             # Add batch normalization, set fused = True for compactness
```
Comment (taylorfturner) [inline dataprofiler/labelers/character_level_cnn_model.py:549]:
```suggestion
            # Add batch normalization
```

Comment (taylorfturner) [reply dataprofiler/labelers/character_level_cnn_model.py:549]: no need for extra part of comment now

File: dataprofiler/labelers/char_load_tf_model.py [hunk 26]
```diff
@@ -294,30 +302,33 @@ def _reconstruct_model(self) -> None:
...
-            self._model.layers.pop()
-
         # Add the final Softmax layer to the previous spot
+        # self._model.layers[-2] to skip: original softmax
         final_softmax_layer = tf.keras.layers.Dense(
             num_labels, activation="softmax", name="softmax_output"
-        )(self._model.layers[-4].output)
+        )(self._model.layers[-2].output)
 
         # Output the model into a .pb file for TensorFlow
```
Comment (taylorfturner) [inline dataprofiler/labelers/char_load_tf_model.py]: out of date comment?

File: dataprofiler/labelers/char_load_tf_model.py [hunk 27]
```diff
@@ -253,20 +254,27 @@ def _construct_model(self) -> None:
             )(self._model.layers[softmax_layer_ind - 1].output)
 
         # Output the model into a .pb file for TensorFlow
```
[the remark this answers is not in the document]
Comment (taylorfturner) [reply dataprofiler/labelers/char_load_tf_model.py]: out of date

File: dataprofiler/labelers/character_level_cnn_model.py (hunk 25 above)
Comment (JGSweets) [reply dataprofiler/labelers/character_level_cnn_model.py]: Just removed it entirely since the line essentially would be the comment

**Review by taylorfturner — DISMISSED** [review]
LGTM 🚀 -- thx, @JGSweets for the contrib!

**Comment by taylorfturner** [comment]
Noticing [this](https://www.tensorflow.org/api_docs/python/tf/keras/*) in the keras 2 -> 3 conversion directions too. Haven't tested fully yet, but could be at issue

**Comment by lettergram** [comment]
Hello, any way we can get this merged?

**Comment by taylorfturner** [comment]
> Hello, any way we can get this merged?

Hey @lettergram, yes. 

- Needs a rebase / resolution of two conflicts now
- I'll be doing a release prior to June 14th. Trying to validate this on my end because the proposed changes hang locally with no error.

Thanks!

**Comment by JGSweets** [comment]
@taylorfturner @lettergram rebase complete!

**Review by taylorfturner — APPROVED** [review]
LGTM

**Review by ksneab7 — APPROVED** [review]
